File: apps/ios/Loonext/Theme/Measures.swift
import SwiftUI

/// #540 — the dashboard's measures, drawn.
///
/// The hand-port of `apps/web/src/components/ui/proportion-ring.tsx` and
/// `share-bar.tsx`, and the third copy after `ui/common/Measures.kt`. The phones
/// shipped the same four measures as numbers in boxes while the laptop got the
/// marks, which reads as two different products — a founder comparing the two
/// sees the same figures presented as if one screen had not been finished.
///
/// ## Two shapes, and the difference is not decorative
///
/// A RING says "how much of one thing" — one part against one whole. A BAR says
/// what a whole is MADE of. The quotes panel asks the second question (won, still
/// out, gone quiet), and forcing that into an arc loses the middle one, which is
/// the only one anybody can still act on.
///
/// ## Accessibility is the part these usually get wrong
///
/// A shape alone is nothing to VoiceOver and nothing to somebody who cannot
/// separate the tones. Both take a SENTENCE from the caller, and neither ever
/// carries a figure that is not also written on the card as text.
///
/// Colour comes from the CALLER: a colour is a fill or a label, never both
/// (D100), and only the card knows which of its own tones mean what.

/// A proportion, as a closing ring.
///
/// Nothing done draws NO arc rather than a dot — a round cap at zero length still
/// paints a mark, and a mark reads as a small amount of something rather than
/// none of it. A caller reporting more done than exists gets a closed ring rather
/// than an arc that has wrapped round and looks like almost nothing.
struct ProportionRing: View {
    let value: Double
    let total: Double
    /// What VoiceOver says. A sentence, not a percentage.
    let label: String
    let color: Color
    var size: CGFloat = 22

    private var fraction: Double {
        let safeTotal = max(0, total)
        guard safeTotal > 0 else { return 0 }
        return min(max(0, value), safeTotal) / safeTotal
    }

    private var stroke: CGFloat { max(2.5, size / 9) }

    var body: some View {
        ZStack {
            // The track. Deliberately faint: it is the amount still to do, and a
            // strong ring for the part NOT done reads as a warning about work
            // that may be perfectly fine.
            Circle()
                .stroke(color.opacity(0.15), lineWidth: stroke)
            if fraction > 0 {
                Circle()
                    .trim(from: 0, to: fraction)
                    .stroke(
                        color,
                        style: StrokeStyle(lineWidth: stroke, lineCap: .round)
                    )
                    // From the top rather than from three o'clock, which is where
                    // every reader expects a progress ring to start.
                    .rotationEffect(.degrees(-90))
            }
        }
        .padding(stroke / 2)
        .frame(width: size, height: size)
        .accessibilityElement()
        .accessibilityLabel(label)
    }
}

/// One part of a whole, for ``ShareBar``.
struct ShareSegment: Identifiable {
    let label: String
    let value: Double
    let color: Color

    var id: String { label }
}

/// A whole, split into its parts.
///
/// Segments summing to LESS than the total leave the remainder as bare track,
/// which is the honest picture — the gap is the part nobody has accounted for,
/// and stretching the parts to fill the bar would hide the number worth chasing.
struct ShareBar: View {
    let segments: [ShareSegment]
    let total: Double
    /// What VoiceOver says. A sentence, not a set of percentages.
    let label: String
    var height: CGFloat = 6

    /// Clamped cumulatively, so a caller whose parts add to more than the whole
    /// gets a full bar rather than segments running off the end. That happens for
    /// real: the parts and the total are separate figures from the server, and a
    /// lagging window can disagree with itself by one.
    private var shares: [(segment: ShareSegment, fraction: Double)] {
        let safeTotal = max(0, total)
        guard safeTotal > 0 else { return [] }
        var used = 0.0
        return segments.map { segment in
            let v = min(max(0, segment.value), safeTotal - used)
            used += v
            return (segment, v / safeTotal)
        }
    }

    var body: some View {
        // Nothing to divide. An empty track reads as a panel that failed to load
        // rather than as a month with no quotes in it.
        if max(0, total) == 0 {
            EmptyView()
        } else {
            GeometryReader { geo in
                HStack(spacing: 0) {
                    ForEach(shares, id: \.segment.id) { share in
                        if share.fraction > 0 {
                            Rectangle()
                                .fill(share.segment.color)
                                .frame(width: geo.size.width * share.fraction)
                        }
                    }
                    Spacer(minLength: 0)
                }
            }
            .frame(height: height)
            // The track BEHIND the segments, so the unaccounted remainder is
            // visible as a gap rather than as nothing at all.
            .background(BrandColor.inset)
            .clipShape(Capsule())
            .accessibilityElement()
            .accessibilityLabel(label)
        }
    }
}

/// The small heading above a measure card.
///
/// #540: two of the four measures put their title INSIDE the card and two put it
/// above, so the four read as two different species of panel in one list. On web
/// the same split showed up as card tops thirty pixels apart in a row, which is a
/// large part of what "looks amateur" meant on that screen.
///
/// One heading, so there is one answer. `trailing` is for the 7/30/90 window the
/// two windowed cards carry; the others pass a quiet note or nothing.
struct MeasureHeader<Trailing: View>: View {
    let label: String
    @ViewBuilder var trailing: () -> Trailing

    var body: some View {
        HStack(alignment: .firstTextBaseline, spacing: 6) {
            Text(label.uppercased())
                .font(.golos(10.5, weight: .bold))
                .kerning(1.2)
                .foregroundStyle(BrandColor.muted500)
            Spacer(minLength: 8)
            trailing()
        }
        .padding(.horizontal, 6)
        .padding(.bottom, 7)
    }
}

extension MeasureHeader where Trailing == EmptyView {
    init(_ label: String) {
        self.label = label
        self.trailing = { EmptyView() }
    }
}

File: apps/android/app/src/main/kotlin/com/loonext/android/ui/common/Measures.kt
package com.loonext.android.ui.common

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.layout.layout
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * #540 — the dashboard's measures, drawn.
 *
 * The hand-port of `apps/web/src/components/ui/proportion-ring.tsx` and
 * `share-bar.tsx`. The phones shipped the same four measures as numbers in boxes
 * while the laptop got the marks, which is the version of this issue that reads
 * as two different products — a founder comparing the two sees the same figures
 * presented as if one of the screens had not been finished.
 *
 * ## Two shapes, and the difference is not decorative
 *
 * A RING says "how much of one thing" — one part against one whole. A BAR says
 * what a whole is MADE of. The quotes panel asks the second question (won, still
 * out, gone quiet), and forcing that into an arc loses the middle one, which is
 * the only one anybody can still act on.
 *
 * ## Accessibility is the part these usually get wrong
 *
 * A shape alone is nothing to TalkBack and nothing to somebody who cannot
 * separate the tones. Both take a SENTENCE from the caller, and neither ever
 * carries a figure that is not also written on the card as text.
 *
 * Colour comes from the CALLER, not from a palette chosen here: a colour is a
 * fill or a label, never both (D100), and only the card knows which of its own
 * tones mean what.
 */

/**
 * A proportion, as a closing ring.
 *
 * Nothing done draws NO arc rather than a dot — a round cap at zero length still
 * paints a mark, and a mark reads as a small amount of something rather than none
 * of it. A caller reporting more done than exists gets a closed ring rather than
 * an arc that has wrapped round and looks like almost nothing.
 */
@Composable
fun ProportionRing(
    value: Float,
    total: Float,
    /** What TalkBack says. A sentence, not a percentage. */
    label: String,
    color: Color,
    modifier: Modifier = Modifier,
    size: Dp = 22.dp,
) {
    val safeTotal = total.coerceAtLeast(0f)
    val safeValue = value.coerceIn(0f, safeTotal)
    val fraction = if (safeTotal == 0f) 0f else safeValue / safeTotal
    val strokeDp = (size.value / 9f).coerceAtLeast(2.5f).dp

    Canvas(
        modifier
            .size(size)
            .semantics { contentDescription = label },
    ) {
        val stroke = strokeDp.toPx()
        val inset = stroke / 2f
        val arcSize = Size(this.size.width - stroke, this.size.height - stroke)
        // The track. Deliberately faint: it is the amount still to do, and a
        // strong ring for the part NOT done reads as a warning about work that
        // may be perfectly fine.
        drawArc(
            color = color.copy(alpha = 0.15f),
            startAngle = 0f,
            sweepAngle = 360f,
            useCenter = false,
            topLeft = Offset(inset, inset),
            size = arcSize,
            style = Stroke(width = stroke),
        )
        if (fraction > 0f) {
            drawArc(
                color = color,
                // From the top rather than from three o'clock, which is where
                // every reader expects a progress ring to start.
                startAngle = -90f,
                sweepAngle = 360f * fraction,
                useCenter = false,
                topLeft = Offset(inset, inset),
                size = arcSize,
                style = Stroke(width = stroke, cap = StrokeCap.Round),
            )
        }
    }
}

/** One part of a whole, for [ShareBar]. */
data class ShareSegment(val label: String, val value: Float, val color: Color)

/**
 * What share of the bar each part gets.
 *
 * PULLED OUT AS A PURE FUNCTION ON PURPOSE. This was first checked by rendering a
 * bar whose parts exceeded its whole and asserting the bar appeared — which it
 * did, clamp or no clamp, because an over-long segment overflows quietly rather
 * than failing. That test passed with the clamp deleted, so it was checking
 * nothing. Arithmetic is testable as arithmetic; rendering is testable as
 * rendering; and conflating them is how a guard ends up decorative.
 *
 * Clamped cumulatively, so a caller whose parts add to more than the whole gets a
 * full bar rather than segments running off the end. That happens for real: the
 * parts and the total are separate figures from the server, and a lagging window
 * can disagree with itself by one.
 */
internal fun shareFractions(values: List<Float>, total: Float): List<Float> {
    val safeTotal = total.coerceAtLeast(0f)
    if (safeTotal == 0f) return values.map { 0f }
    var used = 0f
    return values.map { value ->
        val v = value.coerceIn(0f, safeTotal - used)
        used += v
        v / safeTotal
    }
}

/**
 * A whole, split into its parts.
 *
 * Segments summing to LESS than the total leave the remainder as bare track,
 * which is the honest picture — the gap is the part nobody has accounted for, and
 * stretching the parts to fill the bar would hide the number worth chasing.
 */
@Composable
fun ShareBar(
    segments: List<ShareSegment>,
    total: Float,
    /** What TalkBack says. A sentence, not a set of percentages. */
    label: String,
    modifier: Modifier = Modifier,
    height: Dp = 6.dp,
) {
    val safeTotal = total.coerceAtLeast(0f)
    // Nothing to divide. An empty track reads as a panel that failed to load
    // rather than as a month with no quotes in it.
    if (safeTotal == 0f) return

    val drawn = segments.zip(shareFractions(segments.map { it.value }, safeTotal))

    Row(
        modifier
            .fillMaxWidth()
            .height(height)
            .clip(RoundedCornerShape(percent = 50))
            // The track BEHIND the segments, so the unaccounted remainder is
            // visible as a gap rather than as nothing at all.
            .background(MaterialTheme.colorScheme.surfaceVariant)
            .semantics { contentDescription = label },
    ) {
        drawn.forEach { (segment, fraction) ->
            if (fraction > 0f) {
                Canvas(
                    Modifier
                        .fillMaxHeight()
                        // `weight` cannot express "this share of the parent" when
                        // the remainder must stay empty, so the width is measured
                        // against the incoming constraint directly.
                        .layout { measurable, constraints ->
                            val w = (constraints.maxWidth * fraction).toInt()
                            val placeable = measurable.measure(
                                constraints.copy(minWidth = w, maxWidth = w),
                            )
                            layout(w, placeable.height) { placeable.place(0, 0) }
                        },
                ) {
                    drawRect(color = segment.color)
                }
            }
        }
    }
}

